bindings/go/src/fdb/fdb.go

```diff
@@ -195,48 +194,74 @@ func MustGetAPIVersion() int {
 var apiVersion int
 var networkStarted bool
 var networkMutex sync.RWMutex
+var networkRunning sync.WaitGroup
 
 var openDatabases sync.Map
 
-func startNetwork() error {
+// ensureNetworkIsStarted ensures that the API version is set and then starts the internal network event loop, if not already done.
+// The provided function is guaranteed to be run while network thread is running.
+func ensureNetworkIsStarted(f func()) error {
 	networkMutex.RLock()
 	if networkStarted {
+
+		// network thread is guaranteed to be running while this user-provided function runs
+		f()
+
 		networkMutex.RUnlock()
 		return nil
 	}
 	// release read lock and acquire write lock
 	networkMutex.RUnlock()
 	networkMutex.Lock()
 	defer networkMutex.Unlock()
-	if networkStarted {
-		return nil
-	}
 
-	if e := C.fdb_setup_network(); e != 0 {
-		return Error{int(e)}
-	}
-
-	go func() {
-		e := C.fdb_run_network()
-		if e != 0 {
-			log.Printf("Unhandled error in FoundationDB network thread: %v (%v)\n", C.GoString(C.fdb_get_error(e)), e)
+	// check if meanwhile another goroutine started the network thread
+	if !networkStarted {
+		if e := C.fdb_setup_network(); e != 0 {
+			return Error{int(e)}
 		}
-	}()
 
-	networkStarted = true
+		networkRunning.Add(1)
+		go func() {
+			e := C.fdb_run_network()
+			networkRunning.Done()
+			if e != 0 {
+				panic(fmt.Sprintf("Unhandled error in FoundationDB network thread: %v (%v)\n", C.GoString(C.fdb_get_error(e)), e))
+			}
+		}()
+
+		networkStarted = true
+	}
+
+	// network thread is guaranteed to be running while this user-provided function runs
+	f()
 
 	return nil
 }
 
 // Deprecated: the network is started automatically when a database is opened.
-// StartNetwork initializes the FoundationDB client networking engine. StartNetwork
-// must not be called more than once.
+// StartNetwork does nothing, but it will ensure that the API version is set and return an error otherwise.
 func StartNetwork() error {
 	if apiVersion == 0 {
 		return errAPIVersionUnset
 	}
 
-	return startNetwork()
+	return nil
+}
+
+// StopNetwork signals the internal network event loop to terminate and waits for its termination.
+// This function does nothing if the network has not yet started.
+// Once network is stopped it cannot be started again.
+// See also: https://github.com/apple/foundationdb/issues/3015
+func StopNetwork() {
```

Comment:
```suggestion
func StopNetwork() error {
```

We probably should return an error if the network was started and was already stopped? I'm not sure what will happen, but a user-friendly error like "network was already stopped" might be better than having the code panic or return nothing.

Reply:
Sure, but what about the case of network not yet started? Should it return an error or do nothing?

Reply:
I made it return an error also in such case.